patches/server/0066-lithium-cache-chunk-gen-sea-level.patch

```diff
@@ -0,0 +1,47 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: SuperCoder7979 <25208576+SuperCoder7979@users.noreply.github.com>
+Date: Fri, 22 Jan 2021 16:38:19 -0500
+Subject: [PATCH] lithium: cache chunk gen sea level
+
+Chunk generator settings are passed to the noise chunk generator through a supplier, which retrieves a given chunk generator settings registry key from the registry. The problem arises as this setting is retrieved from the supplier to get the sea level, which is called every single time a block is placed by the generator, for a total of 65536 registry lookups per chunk. As you can imagine this isn't all that fast and removing it speeds up the chunk generation time by a moderate amount (13% of CPU time spent in populateNoise without the patch vs 10% with the patch). My solution to this was to stick the supplier into a Lazy<> so it's computed once and then cached, but I'm open to suggestions. (The sea level can't be computed ahead of time due to a quirk in the new registry system, which would have been the best option here.)
+
+This code was originally made by SuperCoder7979 in a Pull Request to lithium (https://github.com/jellysquid3/lithium-fabric/pull/179)
+
+This code was orignally licensed under the LGPL-3.0 license.
+
+diff --git a/src/main/java/net/minecraft/server/ChunkGeneratorAbstract.java b/src/main/java/net/minecraft/server/ChunkGeneratorAbstract.java
+index fa60285c0c48147ad09b9197bfe577f504dd34bd..9cff4bcff7ac9211ca230fe7ce49ffd4e8ab4035 100644
+--- a/src/main/java/net/minecraft/server/ChunkGeneratorAbstract.java
++++ b/src/main/java/net/minecraft/server/ChunkGeneratorAbstract.java
+@@ -17,6 +17,7 @@ import javax.annotation.Nullable;
+ 
+ public final class ChunkGeneratorAbstract extends ChunkGenerator {
+ 
++    private int cachedSeaLevel = Integer.MIN_VALUE; // Yatopia - lithium cache chunk gen settings
+     public static final Codec<ChunkGeneratorAbstract> d = RecordCodecBuilder.create((instance) -> {
+         return instance.group(WorldChunkManager.a.fieldOf("biome_source").forGetter((chunkgeneratorabstract) -> {
+             return chunkgeneratorabstract.b;
+@@ -688,10 +689,22 @@ public final class ChunkGeneratorAbstract extends ChunkGenerator {
+         return this.x;
+     }
+ 
++    // Yatopia start - lithium cache chunk gen settings
++    /**
++     * Use cached sea level instead of retrieving from the registry every time.
++     * This method is called for every block in the chunk so this will save a lot of registry lookups.
++     *
++     * @author SuperCoder79
++     */
+     @Override
+     public int getSeaLevel() {
+-        return ((GeneratorSettingBase) this.h.get()).g();
++        if (this.cachedSeaLevel == Integer.MIN_VALUE) {
++            this.cachedSeaLevel = ((GeneratorSettingBase) this.h.get()).g();
```

Comment:
obfhelpers

Reply:
Fixed

Reply:
I did the obfhelper wrong lol. I'll fix it shortly.

Reply:
Turns out adding an obfhelper in `GeneratorSettingBase.java` screws up everything, so I'm just not going to add one lol. (see #developer-general in discord for more details)